File: clients/kaya-go/retry_test.go
package kaya

import (
	"testing"
	"time"
)

func TestBackoffExponentialAndCapped(t *testing.T) {
	policy := RetryPolicy{
		MaxAttempts: 5,
		BaseBackoff: 100 * time.Millisecond,
		MaxBackoff:  400 * time.Millisecond,
		Jitter:      false,
	}
	var s uint64 = 1
	if got := policy.Backoff(0, &s); got != 100*time.Millisecond {
		t.Fatalf("backoff(0) = %v, want 100ms", got)
	}
	if got := policy.Backoff(1, &s); got != 200*time.Millisecond {
		t.Fatalf("backoff(1) = %v, want 200ms", got)
	}
	if got := policy.Backoff(2, &s); got != 400*time.Millisecond {
		t.Fatalf("backoff(2) = %v, want 400ms", got)
	}
	// Capped.
	if got := policy.Backoff(3, &s); got != 400*time.Millisecond {
		t.Fatalf("backoff(3) = %v, want 400ms (capped)", got)
	}
	if got := policy.Backoff(30, &s); got != 400*time.Millisecond {
		t.Fatalf("backoff(30) = %v, want 400ms (capped)", got)
	}
}

func TestJitterWithinBoundsAndDeterministic(t *testing.T) {
	policy := RetryPolicy{
		MaxAttempts: 4,
		BaseBackoff: 100 * time.Millisecond,
		MaxBackoff:  1000 * time.Millisecond,
		Jitter:      true,
	}
	var s1 uint64 = 42
	var s2 uint64 = 42
	for retry := uint32(0); retry < 6; retry++ {
		a := policy.Backoff(retry, &s1)
		b := policy.Backoff(retry, &s2)
		if a != b {
			t.Fatalf("same seed must yield same jitter: retry=%d a=%v b=%v", retry, a, b)
		}
		capMs := uint64(100) << retry
		if capMs > 1000 {
			capMs = 1000
		}
		if uint64(a.Milliseconds()) > capMs {
			t.Fatalf("jitter exceeded cap: got %v cap %dms", a, capMs)
		}
	}
}

func TestRetryPolicyNoneIsSingleShot(t *testing.T) {
	p := RetryPolicyNone()
	if p.MaxAttempts != 1 {
		t.Fatalf("max attempts = %d, want 1", p.MaxAttempts)
	}
	if p.RequestTimeout != 0 {
		t.Fatalf("request timeout = %v, want 0", p.RequestTimeout)
	}
}

func TestDefaultRetryPolicy(t *testing.T) {
	p := DefaultRetryPolicy()
	if p.MaxAttempts != 4 {
		t.Fatalf("max attempts = %d, want 4", p.MaxAttempts)
	}
	if p.BaseBackoff != 50*time.Millisecond {
		t.Fatalf("base backoff = %v, want 50ms", p.BaseBackoff)
	}
	if p.MaxBackoff != 2*time.Second {
		t.Fatalf("max backoff = %v, want 2s", p.MaxBackoff)
	}
	if !p.Jitter {
		t.Fatal("jitter should be true by default")
	}
	if p.RequestTimeout != 5*time.Second {
		t.Fatalf("request timeout = %v, want 5s", p.RequestTimeout)
	}
}

func TestSeedFromAddrNonZero(t *testing.T) {
	a := seedFromAddr("127.0.0.1:7379")
	b := seedFromAddr("127.0.0.1:7380")
	if a == 0 || b == 0 {
		t.Fatal("seed must be non-zero")
	}
	if a == b {
		t.Fatal("different addrs should produce different seeds")
	}
	if a&1 == 0 {
		t.Fatal("seed must have low bit set")
	}
}

File: clients/kaya-ts/test/retry.test.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  backoffMs,
  defaultRetryPolicy,
  retryPolicyNone,
  seedFromAddr,
  type RetryPolicy,
  type RngState,
} from "../src/retry.ts";

describe("RetryPolicy", () => {
  it("default matches Go/Rust (4 attempts, 50ms, 2s, jitter, 5s timeout)", () => {
    const p = defaultRetryPolicy();
    assert.equal(p.maxAttempts, 4);
    assert.equal(p.baseBackoffMs, 50);
    assert.equal(p.maxBackoffMs, 2_000);
    assert.equal(p.jitter, true);
    assert.equal(p.requestTimeoutMs, 5_000);
  });

  it("none is single-shot with no timeout", () => {
    const p = retryPolicyNone();
    assert.equal(p.maxAttempts, 1);
    assert.equal(p.requestTimeoutMs, 0);
  });

  it("backoff is exponential and capped", () => {
    const policy: RetryPolicy = {
      maxAttempts: 5,
      baseBackoffMs: 100,
      maxBackoffMs: 400,
      jitter: false,
      requestTimeoutMs: 0,
    };
    const rng: RngState = { seed: 1n };
    assert.equal(backoffMs(policy, 0, rng), 100);
    assert.equal(backoffMs(policy, 1, rng), 200);
    assert.equal(backoffMs(policy, 2, rng), 400);
    assert.equal(backoffMs(policy, 3, rng), 400);
    assert.equal(backoffMs(policy, 30, rng), 400);
  });

  it("jitter stays within bounds and is deterministic", () => {
    const policy: RetryPolicy = {
      maxAttempts: 4,
      baseBackoffMs: 100,
      maxBackoffMs: 1000,
      jitter: true,
      requestTimeoutMs: 0,
    };
    const s1: RngState = { seed: 42n };
    const s2: RngState = { seed: 42n };
    for (let retry = 0; retry < 6; retry++) {
      const a = backoffMs(policy, retry, s1);
      const b = backoffMs(policy, retry, s2);
      assert.equal(a, b, `same seed must yield same jitter at retry=${retry}`);
      const cap = Math.min(100 * 2 ** retry, 1000);
      assert.ok(a <= cap, `jitter exceeded cap: got ${a} cap ${cap}`);
      assert.ok(a >= 0);
    }
  });

  it("seedFromAddr is non-zero, odd, and addr-sensitive", () => {
    const a = seedFromAddr("127.0.0.1:7379");
    const b = seedFromAddr("127.0.0.1:7380");
    assert.notEqual(a, 0n);
    assert.notEqual(b, 0n);
    assert.notEqual(a, b);
    assert.equal(a & 1n, 1n);
  });
});
